### PR TITLE
Fix bindables not being returned by OsuScreen soon enough

### PR DESCRIPTION
**Pull request by nyquillerium**
Closes https://github.com/ppy/osu/issues/4377

### CODE REVIEW

**Comment by peppy** [comment]
Have you tested this is a screen that calls `Expire()` in `OnExiting`? Depending on order of execution there is a chance the schedule will not run.

**Comment by nyquillerium** [comment]
This now changes all instances of ScreenStack being created to use the new OsuScreenStack solution-wide. I have left parameter types for now, could potentially change those as well once we confirm this is what we want.

**Comment by peppy** [comment]
Please at some comments to `OsuScreenStack`.

"Although UnbindAll is automatically called in disposal logic, this is too late as leases can be in leased states while screens are still being transitioned" or something like that.

When adding new special cases, make sure to explain them. If such comments don't exist, someone will come along and be like "hey this isn't actually needed!".

I'd also recommend adding a test to ensure the above, so if someone did remove the logic it would at least fail that.

File: osu.Game/Screens/OsuScreenStack.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        public OsuScreenStack(IScreen baseScreen)
+            : base(baseScreen)
+        {
+            ScreenExited += unbindAllDependencies;
+        }
+
+        /// <summary> Return all child dependency bindables created by the exiting screen </summary>
+        /// <remarks>
+        /// While all bindables will eventually be returned by disposal logic,
+        /// Bindables that are created by every OsuScreen, namely ones created in <see cref="OsuScreenDependencies"/>, will not be returned in time.
```
Comment (peppy) [inline osu.Game/Screens/OsuScreenStack.cs]: this comment is wrong

**Comment by peppy** [comment]
Moved unbind method to `OsuScreen`, updated comments slightly.

**Review by peppy — APPROVED** [review]

File: osu.Game/Screens/OsuScreen.cs [hunk 2]
```diff
@@ -73,6 +73,20 @@ protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnl
...
 
+        /// <summary>
+        /// Unbind and return leases for all <see cref="Bindable{T}"/>s managed by the exiting screen.
+        /// </summary>
+        /// <remarks>
+        /// While all bindables will eventually be returned by disposal logic, this is too late as
+        /// leases would be in a leased state during exiting transitions.
+        /// This localised handling ensures a correct leased state before <see cref="IScreen.OnResuming"/>.
+        /// </remarks>
+        public void UnbindAll()
```
Comment (smoogipoo) [inline osu.Game/Screens/OsuScreen.cs]: `virtual`.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/OsuScreen.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/OsuScreen.cs]: On further discussion, we agreed to move this logic completely to the framework side (`ScreenStack` should call the internal `UnbindAllBindables`).

**Comment by peppy** [comment]
Will be reimplemented on framework side.

**Pull request closed without merging**